tiledb/sm/query/reader.cc

```diff
@@ -2634,21 +2607,56 @@ Status Reader::copy_attribute_values(
   }
 
   // Pre-load all attribute offsets into memory.
-  load_offsets(names);
+  load_tile_offsets(names);
 
-  // Read, unfilter, and copy all attributes into memory.
+  // Instantiate context caches for copying fixed and variable
+  // cells.
   CopyFixedCellsContextCache fixed_ctx_cache;
   CopyVarCellsContextCache var_ctx_cache;
-  for (const auto& name : names) {
-    RETURN_CANCEL_OR_ERROR(read_tiles(name, result_tiles));
-    RETURN_CANCEL_OR_ERROR(unfilter_tiles(name, result_tiles, &cs_ranges));
-    if (!array_schema_->var_size(name))
-      RETURN_CANCEL_OR_ERROR(
-          copy_fixed_cells(name, stride, result_cell_slabs, &fixed_ctx_cache));
-    else
+
+  // The maximum number of attribute to read and unfilter in parallel.
+  // Each attribute requires additional memory to buffer reads into
+  // before copying them back into `buffers_`. Cells must be copied
+  // before moving onto the next set of concurrent reads to prevent
+  // bloating memory. Additionally, the copy cells paths are performed
+  // in serial, which will bottleneck the read concurrency. Increasing
+  // this number will have diminishing returns on performance.
+  const uint64_t concurrent_reads = 2;
```

Comment:
Perhaps make this a constant and add it to `misc/constants.{h,cc}`? We do not need to expose it to the user just yet, but it may be worth gathering all tweaks across the code base in the constants files for internal use.

Reply:
Done. Eventually to be replaced by a memory budget, though.